Automattic/simplenote-ios
# Interlinks Autocomplete: Mark II

### Details
In this PR we're implementing the core logic of Interlinks:

- **InterlinkProcessor:** In charge of determining when the Interlink Autocomplete must be shown / hidden
- **InterlinkViewController:** Renders Autocomplete suggestions above / below the cursor

**Note:** Nothing is expected to happen, at this stage, when you press over any row.

**Apologies about the size of the PR!!!** I've tried splitting this further, but testing without an actual UI makes no sense.

cc @eshurakov **Thank yoou sir!!!**

Ref. #914

---

### Test: Orientation
1. Open the Editor and add a note with 6 lines
2. Place the cursor in the 6th line
3. Type `[keyword` (with `keyword` being a word contained in any of your note's titles)

- [x] Verify Autocomplete shows up below the cursor's location
- [x] Verify that scrolling over the TextView below causes the Autocomplete window to follow along
- [x] Verify that if you scroll enough, the Autocomplete UI shows **below the Navigation Bar**
- [x] Verify that if your (updated) results yield just one row, the Autocomplete View doesn't jump above the keyword
- [x] Verify that if there are (way more) lines above, typing `[keyword` displays Autocomplete **above** the cursor

### Test: Filtering
1. Add a new note
2. In the new note's title, enter a word that's not present anywhere else. Suggestion `TEST999` perhaps?
3. Place the cursor in a newline
4. Type `[TEST999`

- [x] Verify Autocomplete does not show up (we intend to prevent circular references!)


### Test: Adaptivity
1. Open the Editor and type `[keyword` to trigger Autocomplete

- [x] Verify that rotating the device causes Autocomplete to disappear / reappear in the right location

### Test: UX
1. Open the Editor and type `[keyword` to trigger Autocomplete

- [x] Verify that a maximum of `3.5` rows always show up (we intend to avoid huge tables!!)
- [x] Verify Autocomplete reacts well to Dark / Light mode
- [ ] Verify Autocomplete reacts well to Dynamic Fonts
- [x] Verify that, on highlight, Autocomplete's First and Last rows get rounded corners

### Test: Dismissal
Please trigger the Autocomplete window, and verify it can be dismissed as follows:

- [x] Verify that pressing anywhere else in the TextView causes Autocomplete to get dismissed
- [x] Verify that pressing over any of the Navigation Bar buttons causes interlink to dismiss
- [x] Verify that pressing Back causes Interlink to get dismissed

### Release
These changes do not require release notes.

### Screenshots

<img width="200" src="https://user-images.githubusercontent.com/1195260/98828747-352e3300-2417-11eb-8aef-b8c276ed295e.png"> <img width="200" src="https://user-images.githubusercontent.com/1195260/98828764-38c1ba00-2417-11eb-8989-2e36be0312e6.png">
